### PR TITLE
ENH: allow storing ExtensionArrays in the Index

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
I started experimenting with allowing EAs to be used in the Index (so we can put nullable int/bool, string dtype etc in the Index).

My approach is to use the base `Index` class for this (and not a custom subclass):

- We use a similar approach for Series, which can also hold either ndarray or EA as underlying values
- I don't think we want to start adding subclasses for all new EAs (like Int64Index for the new Int64Dtype is not quite possible ..)
- This also works for external EAs

For now, it just integrates it, so an EA can be put in an index and tested operations work. This PR does *not* (yet) handle ensuring that operations relying on the index' Engine are efficient (currently it simply uses `np.asarray(EA)` as the engine values, which eg will be object dtype for Int64Dtype)

cc @jbrockmendel @TomAugspurger @jreback

### CODE REVIEW

**Comment by jbrockmendel** [comment]
This is the idea behind ExtensionIndex

**Comment by jreback** [comment]
I agree with @jbrockmendel here, this is the exactly the point of ExtensionIndex. trying to modify Index doesn't make much sense.

**Comment by jorisvandenbossche** [comment]
Sorry, I should have been more explicit regarding ExtensionIndex in my reasoning above: 

- Yes, this is (somewhat/fully) duplicating `ExtensionIndex`. But I think some things in ExtensionIndex can just be moved to the base class index. Again, for Series we have exactly the same approach of having both ndarray or EA as underlying values (we don't have an "ExtensionSeries").
- Currently, the `ExtensionIndex` is only used as base class that concrete Index classes subclass from. Here, it is meant to be the actual class that users use/see. I personally don't think this should be "ExtensionIndex"

**Comment by jorisvandenbossche** [comment]
> This is the idea behind ExtensionIndex

@jbrockmendel to make it concrete: are there (technical) reasons for you that are specific to Index and that make it more important to use a custom ExtensionIndex subclass instead of handling this in the base Index class?

For example, both have a `__getitem__`, which in both cases dispatch to `self._data[..]`, but with some different checks (eg the base Index also checks for object boolean masks). But is there a fundamental reason those can't be merged?

Other example: the `dropna` in both is exactly the same, except for the use of `self._data` vs `self._values`, but `self._values` is defined to be `self._data`. 
For `repeat` the same. 
For `take` as well, the base class method just has a branch for `_can_hold_na` being False as well (which is never the case for the ExtensionIndex)

(a pre-cursor PR to this one could be to remove the duplicated methods between Index and ExtensionIndex)

**Comment by jbrockmendel** [comment]
> to make it concrete: are there (technical) reasons for you that are specific to Index and that make it more important to use a custom ExtensionIndex subclass instead of handling this in the base Index class?

Nothing that I have a particularly strong opinion about.  If the implementation in this PR turns out to be elegant (unless asked, ill wait to review in detail until green), i wouldn't have a problem with it.

That said, ATM whenever we have an Index object we know it is object-dtype.  I expect there are places in the code where we are relying on that.  (might be worth making an ObjectIndex?  orthogonal discussion)

**Comment by jreback** [comment]
@jorisvandenbossche yeah this is almost certainly not going to 100% overlap with an object dtyped Index base, because that serves 2 master right now, the object dtyped and a base index. If we would create a ObjectIndex then maybe what you are doing would work, but I find it easier / more compatible to simply have an ExtensionArrayIndex that can hold any type of EA.

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -5212,9 +5241,14 @@ def insert(self, loc: int, item):
...
         #  so self is never backed by an EA.
-        arr = np.asarray(self)
         item = self._coerce_scalar_to_index(item)._values
-        idx = np.concatenate((arr[:loc], item, arr[loc:]))
+
+        if is_extension_array_dtype(self.dtype):
+            arr = self._values
+            idx = arr._concat_same_type([arr[:loc], item, arr[loc:]])
+        else:
+            arr = np.asarray(self)
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:5250]: arr = self._values could go outside the if/elif

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -3839,7 +3845,10 @@ def _get_engine_target(self) -> np.ndarray:
         """
         Get the ndarray that we can pass to the IndexEngine constructor.
         """
-        return self._values
+        if isinstance(self._values, np.ndarray):
+            return self._values
+        else:
+            return np.asarray(self._values)
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:3851]: doesnt `np.asarray(self._values)` work unconditionally?

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -346,11 +346,13 @@ def __new__(
                 ea_cls = dtype.construct_array_type()
                 data = ea_cls._from_sequence(data, dtype=dtype, copy=False)
             else:
-                data = np.asarray(data, dtype=object)
+                # TODO clean-up with extract_array ?
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:349]: +1

**Comment by jbrockmendel** [comment]
@jorisvandenbossche im working on getting this feature out the door; i think we can close this to clear the queue

**Comment by jreback** [comment]
closing as duplicate of #37869 can move discussion there.

**Comment by jbrockmendel** [comment]
@jorisvandenbossche I suggested closing this not based on the Index[EA] vs ExtensionIndex difference, but bc you don't seem to be working on it.  If we get the other one working well, we can revisit the possibility of getting Index[EA] working.

(FWIW as I see more and more of the DTI/TDI/PI code getting shared, I've found myself thinking about the possibility those may not need to be subclasses at all [in this daydream, the subclass-specific stuff would be accessed by a .dt accessor xref #17134], which bears a resemblance to the Index[EA] design)

**Comment by jreback** [comment]
@jorisvandenbossche pls update or close

**Comment by jorisvandenbossche** [comment]
We first need to agree on how to do this (see discussion above, and also the overview issue I created at https://github.com/pandas-dev/pandas/issues/39133)

**Comment by mroeschke** [comment]
Appears this PR draft has been dormant for a while so closing. Feel free to reopen when you have time to continue

**Comment by jorisvandenbossche** [comment]
Note for future reference: this work is now actually done in https://github.com/pandas-dev/pandas/pull/43930

**Pull request closed without merging**